plugins/in_systemd/systemd.c

```diff
@@ -276,15 +282,28 @@ static int in_systemd_collect(struct flb_input_instance *i_ins,
     if (ret_j == 0) {
         ctx->pending_records = FLB_FALSE;
         return FLB_SYSTEMD_OK;
-    }
-
+    } else if (ret_j > 0) {
     /*
-     * ret_j == -1, the loop was broken due to some special condition like
+     * ret_j == 1, but the loop was broken due to some special condition like
      * buffer size limit or it reach the max number of rows that it supposed to
      * process on this call. Assume there are pending records.
      */
     ctx->pending_records = FLB_TRUE;
     return FLB_SYSTEMD_MORE;
+    } else {
+        ctx->pending_records = FLB_FALSE;
+        sd_journal_get_cursor(ctx->j, &cursor);
+        tmp = flb_input_get_property("read_from_tail", ctx->i_ins);
```

Comment:
I see that read_from_tail option is being re-read every time if ret_j == -1, which is not optimal (string comparisson) but I found that we didn't have a variable context for that. I've pushed 5a360ec945b72fc1e4d6d6337fb312d875e850bf that adds the new variable so you can do ctx->read_from_trail == FLB_TRUE instead of get the plugin property. Please adjust the code using the new context.

Reply:
Actually it does not makes sense to read from tail in this case: we want to minimize the amount of lost logs, hence we need to read the earliest possible log message, which is done always by sd_journal_seek_head(), irrespective of `read_from_tail` property.